src/schemas/contact.ts

```diff
@@ -41,6 +41,12 @@ export interface ContactPayload {
   signature? : string,
   star?      : boolean,
   weixin?    : string,
+
+  corp?        : string,
+  title?       : string,
+  description? : string,
+  isWeixin?    : boolean,
```

Comment:
Can we use the `type: ContactType` to include those two new types: `Weixin` and `External`?

For example, currently the `ContactType` is:

```ts
export enum ContactType {
  Unknown    = 0,
  Individual = 1,
  Official   = 2,

  /**
   * Huan(202004):
   * @deprecated: use Individual instead
   */
  Personal = Individual,
}
```

Can we use it to carry those new types?

And one more thing is that we need to make it also meaningful for our future non-wechat puppets, for example, puppet-twilio, puppet-mailgun, puppet-weibo, etc.

Reply:
As I described in https://github.com/wechaty/wechaty/issues/2036

The types are in two dimensions:

||internal|external|
|---|---|---|
|weixin|✅|❌|
|work|✅|✅|

I've considered using `ContactType` when I created the PR, then we might need to add three more types to cover all the cases:

```
ExternalWeixin
ExternalWxWork
InternalWxWork
```

If we do this, when we want to filter the contact which are external contact, we need to pick out the contact with type `ExternalWeixin` and `ExternalWxWork`, which is not convenient and is actually adding complexity to developers' code, also for the implementation. So I personally think using separate properties are better choice.

> And one more thing is that we need to make it also meaningful for our future non-wechat puppets, for example, puppet-twilio, puppet-mailgun, puppet-weibo, etc.

I also thought about this issue, the `external` is universal, I believe Dingtalk also has this concept. But for weixin, I can not think of a name that also meaningful for future non-wechat puppets, since the concept is specific for wechat eco-system. Do you have better idea for this?

Reply:
Yes, I understand that we have two dimensions when we are thinking about the WeChat and WXWork together.

However, when there's a wechaty instance running, this instance is very specific using one protocol, for example, it will be either using a WeChat puppet, like puppet-donut or using a WXWork puppet, like puppet-wxwork.

When we are using a specific puppet, then the `external` concept will be very specific:

1. If we are using the WeChat puppet, then the `external` will stand for a non-WeChat contact.
1. If we are using the WXWork puppet, then the `external` will stand for a non-WXWork contact.

If we can accept the design above, then we will be able to use only two types instead of four to describe the internal/external.

Do you think it's a good idea? It's just a proposal for discussion because I'm still thinking and not sure about it.

Reply:
> 1. If we are using the WeChat puppet, then the external will stand for a non-WeChat contact.
> 1. If we are using the WXWork puppet, then the external will stand for a non-WXWork contact.

I think for `external` attribute, this will be unavailable for WeChat puppet, to distinguish WeChat or WxWork contact, we should use `isWeixin` flag, or we can rename it to something like `isCorp`, maybe `isCorp` makes more sense than `isWeixin`.

And for `external` concept, this can only applied to all enterprise IM, like WxWork, Dingtalk or Lark, I believe these IMs has the concept of external and internal.

Reply:
If we want to reuse the `type()` function, maybe we need to create a complex structure, instead of using current enum, say a class called `ContactType`, and we can do something like below:

```ts
enum BaseType {
  Official = 1,
  Weixin = 2,
  Corp = 3,
}
class ContactType {
  public baseType (): BaseType {}
  public external(): boolean {}
}

const type = contact.type()

const isCorp = type.baseType() === BaseType.Corp
const external = type.external()
```

During my typing, I think we can merge the [`weixin` `corp`] concept into the [`official`, `individual`] design, these should be parallel, each contact should have only one type among these types.

Reply:
> During my typing, I think we can merge the [weixin corp] concept into the [official, individual] design, these should be parallel, each contact should have only one type among these types.

Yes, that's absolutely what I'm trying to explain!

Let's keep thinking of this design for a while and try to universal those types.